src/image_undistort.cpp

```diff
@@ -3,59 +3,282 @@
+  if (!process_image_) {
+    camera_info_in_.header.stamp = image_msg_in->header.stamp;
+    camera_info_in_.header.frame_id = image_msg_in->header.frame_id;
+    camera_info_pub_.publish(camera_info_in_);
+  }
+  cv_bridge::CvImageConstPtr image_in_ptr =
+      cv_bridge::toCvShare(image_msg_in, "");
+  cv_bridge::CvImagePtr image_out_ptr(
+      new cv_bridge::CvImage(image_in_ptr->header, image_in_ptr->encoding));
 
   if (undistorter_ptr_) {
-    undistorter_ptr_->undistortImage(image_distorted, &image_undistorted);
+    undistorter_ptr_->undistortImage(image_in_ptr->image,
+                                     &(image_out_ptr->image));
   } else {
     ROS_WARN("Attempted to undistort image before setting valid camera info");
     return;
   }
 
-  cv_bridge::CvImage out_msg;
-  out_msg.header = image_msg->header;
-  out_msg.encoding = image_msg->encoding;
-  out_msg.image = image_undistorted;
+  camera_info_out_.header.stamp = image_out_ptr->header.stamp;
+  camera_info_out_.header.frame_id = image_out_ptr->header.frame_id;
+  camera_pub_.publish(*(image_out_ptr->toImageMsg()), camera_info_out_);
+}
 
-  image_pub_.publish(out_msg);
+void ImageUndistort::cameraCallback(
+    const sensor_msgs::ImageConstPtr& image_msg_in,
+    const sensor_msgs::CameraInfoConstPtr& camera_info_in) {
+  updateCameraInfo(*camera_info_in);
+  imageCallback(image_msg_in);
 }
 
-void ImageUndistort::camInfoCallback(
-    const sensor_msgs::CameraInfoConstPtr& camera_msg) {
-  Eigen::Matrix3d K =
-      Eigen::Map<const Eigen::Matrix<double, 3, 3, Eigen::RowMajor>>(
-          camera_msg->K.elems);
+void ImageUndistort::updateCameraInfo(
+    const sensor_msgs::CameraInfo& camera_info) {
+  // only update if camera parameters have changed
+  if ((camera_info.K == camera_info_in_.K) &&
+      (camera_info.width == camera_info_in_.width) &&
+      (camera_info.height == camera_info_in_.height) &&
+      (camera_info.distortion_model == camera_info_in_.distortion_model) &&
+      (camera_info.D == camera_info_in_.D)) {
+    return;
+  }
+  camera_info_in_ = camera_info;
+  if (!output_camera_info_from_yaml_) {
+    camera_info_out_ = camera_info_in_;
+  }
+
+  for (size_t i = 0; i < camera_info.D.size(); ++i) {
+    if (!undistort_image_) {
+      camera_info_in_.D[i] = 0;
+      camera_info_out_.D[i] = camera_info.D[i];
+    } else {
+      camera_info_in_.D[i] = camera_info.D[i];
+      camera_info_out_.D[i] = 0;
+    }
+  }
 
-  const cv::Size resolution(camera_msg->width, camera_msg->height);
+  Eigen::Map<const Eigen::Matrix<double, 3, 4, Eigen::RowMajor>> P_in(
+      camera_info_in_.P.data());
+
+  Eigen::Map<const Eigen::Matrix<double, 3, 4, Eigen::RowMajor>> P_out(
+      camera_info_out_.P.data());
+
+  cv::Size resolution(camera_info_out_.width, camera_info_out_.height);
 
   bool using_radtan;
-  if (camera_msg->distortion_model == std::string("radtan")) {
+  if ((camera_info.distortion_model == std::string("Plumb Bob")) ||
```

Comment:
Maybe convert to lowercase so it's case independent (I've seen "plumb bob" more often than "Plumb Bob")?

http://math-linux.com/c/faq-c/faq-c-stl/article/how-to-convert-string-to-lower-case-or-upper-case-in-c is one example.